stdlib/public/core/UnfoldSequence.swift

```diff
@@ -0,0 +1,113 @@
+///
+/// This function can be used to replace many cases that were previously handled
+/// using C-style `for` loops.
+///
+/// Example:
+///
+///     // Walk the elements of a tree from a node up to the root
+///     for node in sequence(first: leaf, next: { $0.parent }) {
+///         // node is leaf, then leaf.parent, then leaf.parent.parent, etc.
+///     }
+///
+///     // Iterate over all powers of two (ignoring overflow)
+///     for value in sequence(first: 1, next: { $0 * 2 }) {
+///         // value is 1, then 2, then 4, then 8, etc.
+///     }
+///
+/// - Parameter first: The first value to be returned from the sequence.
+/// - Parameter next: A closure that accepts the previous sequence value and
+///   returns the next value.
+/// - Returns: A sequence that starts with `first` and continues with every
+///   value returned by passing the previous value to `next`.
+///
+/// - SeeAlso: `sequence(state:next:)`
+public func sequence<T>(first: T, next: (T) -> T?) -> UnfoldFirstSequence<T> {
+    // The trivial implementation where the state is the next value to return
+    // has the downside of being unnecessarily eager (it evaluates `next` one
+    // step in advance). We solve this by using a boolean value to disambiguate
+    // between the first value (that's computed in advance) and the rest.
+    return sequence(state: (first, true), next: { (state: inout (T?, Bool)) -> T? in
+        switch state {
+        case (let value, true):
+            state.1 = false
+            return value
+        case (let value?, _):
+            let nextValue = next(value)
+            state.0 = nextValue
+            return nextValue
+        case (nil, _):
+            return nil
+        }
+    })
+}
+
+/// Returns a sequence formed from repeated lazy applications of `next` to a
+/// mutable `state`.
+///
+/// Every value returned by the sequence is the result of invoking `next` with a
+/// mutable state. The same state is passed to all invocations of `next`, so
+/// subsequent calls will see any mutations made by previous calls. The sequence
+/// ends when `next` returns `nil`. If `next` never returns `nil`, the sequence
+/// is infinite.
+///
+/// This function can be used to replace many instances of `AnyIterator` that
+/// wrap a closure.
+///
+/// Example:
+///
+///     // Interleave two sequences that yield the same element type
+///     sequence(state: (false, seq1.makeIterator(), seq2.makeIterator()), next: { iters in
+///         iters.0 = !iters.0
+///         return iters.0 ? iters.1.next() : iters.2.next()
+///     })
+///
+/// - Parameter state: The initial state that will be passed to each closure.
+/// - Parameter next: A closure that accepts an `inout` state and returns the
+///   next element of the sequence.
+/// - Returns: A sequence that yields each successive value from `next.
+///
+/// - SeeAlso: `sequence(first:next:)`
+public func sequence<T, State>(state: State, next: (inout State) -> T?) -> UnfoldSequence<T, State> {
```

Comment:
Regarding `T` => `Element`, I checked before writing this and our functions like `map` still use type parameters that look like `T`. My impression was that we renamed `T` to `Element`, etc when used as parameters to types themselves, but left them as `T` on the functions that return the types, because the type here is not defined as "the element of the sequence", but is instead defined as e.g. "the return value of the closure", and then the resulting sequence uses that return value for its elements.